### PR TITLE
feat: implement bandwidth limit for proxy-cache

### LINKED ISSUE #20589: Enhance proxy-cache feature to limit the bandwidth to pull image from upstream
Enhance proxy-cache feature to limit the bandwidth to pull image from upstream

### PR DESCRIPTION
**Pull request by zyyw**
Thank you for contributing to Harbor!

# Comprehensive Summary of your change

# Issue being fixed
Fixes #20589

Please indicate you've done the following:
- [x] Well Written Title and Summary of the PR
- [x] Label the PR as needed. "release-note/ignore-for-release, release-note/new-feature, release-note/update, release-note/enhancement, release-note/community, release-note/breaking-change, release-note/docs, release-note/infra, release-note/deprecation"
- [x] Accepted the DCO. Commits without the DCO will delay acceptance.
- [x] Made sure tests are passing and test coverage is added if needed.
- [ ] Considered the docs impact and opened a new docs issue or PR with docs changes if needed in [website repository](https://github.com/goharbor/website).

# Testing
test case 1: proxy_speed_kb = "-1"
```
        "metadata": {
            "proxy_speed_kb": "-1",
            "public": "false",
            "retention_id": "1"
        },
        "name": "proxy1-unlimited1",
```
Screenshot 2024-08-06 at 3 13 15 PM

test case 2: proxy_speed_kb = "100"
```
        "metadata": {
            "proxy_speed_kb": "100",
            "public": "false",
            "retention_id": "2"
        },
        "name": "proxy2-limited-100kb",
```
100 is in kilo bit per second, 100/8 = 12.5 kilo bytes per second
Screenshot 2024-08-06 at 3 16 50 PM

test case 3: proxy_speed_kb = "300"
```
        "metadata": {
            "proxy_speed_kb": "300",
            "public": "false",
            "retention_id": "3"
        },
        "name": "proxy3-limited-300kb",
```
300 is in kilo bit per second, 300/8 = 37.5 kilo bytes per second
Screenshot 2024-08-06 at 3 18 08 PM

test case 4: proxy_speed_kb = "1024"
```
        "metadata": {
            "proxy_speed_kb": "1024",
            "public": "false",
            "retention_id": "4"
        },
        "name": "proxy4-limited-1024kb",
```
1024 is in kilo bit per second, 1024/8 = 128 kilo bytes per second
Screenshot 2024-08-06 at 3 23 31 PM

test case 5: proxy_speed_kb = "102400"
```
        "metadata": {
            "proxy_speed_kb": "102400",
            "public": "false",
            "retention_id": "5"
        },
        "name": "proxy5-limited-102400kb",
```
1024 is in kilo bit per second, 1024/8 = 12800 kilo bytes per second
Screenshot 2024-08-06 at 3 23 57 PM

### CODE REVIEW

File: src/server/v2.0/handler/project.go [hunk 1]
```diff
@@ -547,6 +559,11 @@ func (a *projectAPI) UpdateProject(ctx context.Context, params operation.UpdateP
 		}
 	}
 
+	// ignore metadata.proxy_speed_kb for non-proxy-cache project
+	if params.Project.Metadata != nil && !p.IsProxy() {
+		params.Project.Metadata.ProxySpeedKb = nil
+	}
+
```
Comment (yanji09) [inline src/server/v2.0/handler/project.go:566]: Here need to ignore the proxy_speed_kb setting for non proxy cache project. But the code logic here is ignoring the proxy cache project.

File: src/server/v2.0/handler/project_metadata.go [hunk 2]
```diff
@@ -155,6 +155,12 @@ func (p *projectMetadataAPI) validate(metas map[string]string) (map[string]strin
 			return nil, errors.New(nil).WithCode(errors.BadRequestCode).WithMessage("invalid value: %s", value)
 		}
 		metas[proModels.ProMetaSeverity] = strings.ToLower(severity.String())
+	case proModels.ProMetaProxySpeed:
+		v, err := strconv.ParseInt(value, 10, 32)
```
Comment (stonezdj) [inline src/server/v2.0/handler/project_metadata.go:159]: Only proxy cache project can have a speed limit, we should have a check on this

File: src/server/v2.0/handler/project.go (hunk 1 above)
Comment (zyyw) [reply src/server/v2.0/handler/project.go:566]: nice catch! updated.

File: src/server/v2.0/handler/project_metadata.go (hunk 2 above)
Comment (zyyw) [reply src/server/v2.0/handler/project_metadata.go:159]: > Only proxy cache project can have a speed limit, we should have a check on this

this `func (p *projectMetadataAPI) validate() () {}` function merely serves the purpose of checking the metadata key is in the defined list when adding/updating `ProjectMetadatas` disregarding its relationship with project.

Additionally, `params.Project.Metadata.EnableContentTrust` will be set to nil for proxy cache project, while the current logic of `validate()` does not checking proxy-cache either.  

https://github.com/goharbor/harbor/blob/main/src/server/v2.0/handler/project.go#L550-L554

File: api/v2.0/swagger.yaml [hunk 3]
```diff
@@ -7252,6 +7252,10 @@ definitions:
         type: string
         description: 'The ID of the tag retention policy for the project'
         x-nullable: true
+      proxy_speed_kb:
+        type: string
```
Comment (wy65701436) [inline api/v2.0/swagger.yaml:7256]: it should be an integer?

File: src/controller/proxy/remote.go [hunk 4]
```diff
@@ -83,7 +87,14 @@ func (r *remoteHelper) init(ctx context.Context) error {
 }
 
 func (r *remoteHelper) BlobReader(repo, dig string) (int64, io.ReadCloser, error) {
-	return r.registry.PullBlob(repo, dig)
+	sz, bReader, err := r.registry.PullBlob(repo, dig)
+	if err != nil {
+		return 0, nil, err
+	}
+	if r.opts.Speed > 0 {
```
Comment (wy65701436) [inline src/controller/proxy/remote.go:94]:
```
if r.opt !=nil {
  ...
}
```

File: api/v2.0/swagger.yaml [hunk 5]
```diff
@@ -7252,6 +7252,10 @@ definitions:
...
+        description: 'The bandwidth limit of proxy cache, in Kbps (kilobits per second)'
```
Comment (stonezdj) [inline api/v2.0/swagger.yaml:7257]: It limits the communication from Harbor to the upstream registry, not the client to the Harbor.

File: src/server/v2.0/handler/project.go [hunk 6]
```diff
@@ -155,6 +155,18 @@ func (a *projectAPI) CreateProject(ctx context.Context, params operation.CreateP
 		}
 	}
 
+	// ignore metadata.proxy_speed_kb for non-proxy-cache project
+	if req.RegistryID == nil {
+		req.Metadata.ProxySpeedKb = nil
+	}
+
+	// validate metadata.proxy_speed_kb. It should be an int32
+	if ps := req.Metadata.ProxySpeedKb; ps != nil {
```
Comment (wy65701436) [inline src/server/v2.0/handler/project.go:164]: put this into the validateProjectReq()

**Review by wy65701436 — APPROVED** [review]
lgtm

**Review by stonezdj — APPROVED** [review]

**Review by stonezdj — APPROVED** [review]

File: src/controller/proxy/options.go [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package proxy
+
+type Option func(*Options)
+
+type Options struct {
+	// Speed is the data transfer speed for replication, no limit by default.
```
Comment (chlins) [inline src/controller/proxy/options.go:20]: Mismatch comment.

File: src/controller/proxy/remote.go [hunk 8]
```diff
@@ -43,13 +44,16 @@ type remoteHelper struct {
 	regID       int64
 	registry    adapter.ArtifactRegistry
 	registryMgr reg.Manager
+	opts        *Options
 }
 
 // NewRemoteHelper create a remote interface
-func NewRemoteHelper(ctx context.Context, regID int64) (RemoteInterface, error) {
+func NewRemoteHelper(ctx context.Context, regID int64, speed int32) (RemoteInterface, error) {
```
Comment (chlins) [inline src/controller/proxy/remote.go:51]: The use of option closure is for subsequent better scalability, if speed is passed directly as a function parameter then it loses the significance of the option method, a more reasonable use of the following:

```go
func NewRemoteHelper(ctx context.Context, regID int64, opts ...Option) (RemoteInterface, error) {
    o := NewOptions(opts)
}

// call NewRemoteHelper
_, _  := proxy.NewRemoteHelper(ctx, regID, WithSpeed(speed))
```

File: src/controller/proxy/options.go (hunk 7 above)
Comment (zyyw) [reply src/controller/proxy/options.go:20]: nice catch. Thank you!

File: src/controller/proxy/remote.go (hunk 8 above)
Comment (zyyw) [reply src/controller/proxy/remote.go:51]: Good idea!

**Review by chlins — APPROVED** [review]
lgtm

**Pull request merged**